File: ports/go/pipeline/loader.go
package pipeline

import (
	"context"
	"fmt"
	"os"

	"github.com/jagentic/goagentic/core"
	"github.com/jagentic/goagentic/engines/natsjs"
	"gopkg.in/yaml.v3"
)

// Runtime is the backend seam: process one turn.
type Runtime interface {
	Submit(core.Event) core.TurnResult
}

// System is a built, deployed agentic system on the chosen backend.
type System struct {
	BackendName string
	Built       Built
	rt          Runtime
}

// Submit runs a turn on the backend.
func (s *System) Submit(e core.Event) core.TurnResult { return s.rt.Submit(e) }

// Load reads a pipeline.yaml and builds the system (backend from YAML unless overridden).
func Load(path, backend string) (*System, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var spec map[string]any
	if err := yaml.Unmarshal(data, &spec); err != nil {
		return nil, err
	}
	return BuildSystem(spec, backend)
}

// BuildSystem compiles a spec and wires it onto a backend.
func BuildSystem(spec map[string]any, backend string) (*System, error) {
	built, err := Build(spec, defaultChatClient)
	if err != nil {
		return nil, err
	}
	name := backend
	if name == "" {
		name = asString(spec["backend"], "local")
	}
	rt, err := makeBackend(name, built)
	if err != nil {
		return nil, err
	}
	return &System{BackendName: name, Built: built, rt: rt}, nil
}

type localBackend struct{ rt *core.LocalRuntime }

func (b localBackend) Submit(e core.Event) core.TurnResult { return b.rt.Submit(e) }

type natsBackend struct {
	rt  *natsjs.Runtime
	ctx context.Context
}

func (b natsBackend) Submit(e core.Event) core.TurnResult {
	res, err := b.rt.Submit(b.ctx, e)
	if err != nil {
		return core.TurnResult{ConversationID: e.ConversationID, Reply: "[error] " + err.Error(), OK: false}
	}
	return res
}

func makeBackend(name string, built Built) (Runtime, error) {
	switch name {
	case "", "local":
		return localBackend{core.NewLocalRuntime(built.Graph, nil, nil, built.Tools, built.Retriever)}, nil
	case "nats":
		rt := natsjs.New(built.Graph, built.Tools, built.Retriever)
		ctx := context.Background()
		if err := rt.Connect(ctx, os.Getenv("AGENTIC_NATS_URL")); err != nil {
			return nil, fmt.Errorf("nats backend: %w", err)
		}
		return natsBackend{rt: rt, ctx: ctx}, nil
	default:
		return nil, fmt.Errorf("backend %q not wired in Go (have: local, nats; temporal runs via a worker)", name)
	}
}

func defaultChatClient(llm map[string]any) core.ChatClient {
	switch asString(llm["provider"], "ollama") {
	case "ollama":
		return core.NewOllamaChatClient(asString(llm["model"], "qwen2.5:3b"), asString(llm["base_url"], ""))
	case "openai":
		return core.NewOpenAIChatClient(asString(llm["model"], "gpt-5.4-mini"), "", asString(llm["base_url"], ""))
	case "stub":
		var script []core.ChatResult
		for _, raw := range asList(llm["script"]) {
			step := asMap(raw)
			if tool := asString(step["tool"], ""); tool != "" {
				args := map[string]any{}
				for k, v := range asMap(step["args"]) {
					args[k] = v
				}
				script = append(script, core.ToolCall(tool, args))
			} else {
				script = append(script, core.TextResult(asString(step["text"], "ok")))
			}
		}
		if len(script) == 0 {
			script = append(script, core.TextResult("ok"))
		}
		return core.NewStubChatClient(script...)
	default:
		panic("unknown llm provider " + asString(llm["provider"], ""))
	}
}

// ---- spec casting helpers (yaml.v3 gives map[string]any / []any / int / float64) ----

func asMap(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	return map[string]any{}
}

func asList(v any) []any {
	if l, ok := v.([]any); ok {
		return l
	}
	return nil
}

func asString(v any, def string) string {
	if s, ok := v.(string); ok {
		return s
	}
	return def
}

func asInt(v any, def int) int {
	switch n := v.(type) {
	case int:
		return n
	case int64:
		return int(n)
	case float64:
		return int(n)
	}
	return def
}

func asFloat(v any, def float64) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int:
		return float64(n)
	}
	return def
}

func asBool(v any) bool {
	b, _ := v.(bool)
	return b
}

File: ports/agentic-pipeline/agentic_pipeline/loader.py
"""YAML pipeline loader — parse a ``pipeline.yaml``, build the agentic system with the
core GraphBuilder, and select the backend. The result is a runnable system: an object
with ``submit(Event) -> TurnResult`` on the chosen backend.

    from agentic_pipeline.loader import load
    system = load("banking.yaml")          # backend chosen by the YAML
    print(system.submit(Event("c1", "what is my balance?")).reply)
"""

from __future__ import annotations

import sys
from pathlib import Path
from typing import Any, Dict, Optional

import yaml

_PORTS = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(_PORTS / "pyagentic"))

from pyagentic import builder  # noqa: E402
from pyagentic.core import Event  # noqa: E402

from . import backends  # noqa: E402


def _chat_client_factory(llm_spec: Dict[str, Any]):
    """Build a ChatClient from the ``llm:`` section. ``stub`` needs an inline ``script``
    of {tool|text, args} steps (handy for offline tests / deterministic demos)."""
    from pyagentic.llm import ChatResult, OllamaChatClient, OpenAIChatClient, StubChatClient

    provider = (llm_spec or {}).get("provider", "ollama")
    if provider == "ollama":
        return OllamaChatClient(model=llm_spec.get("model", "qwen2.5:3b"), base_url=llm_spec.get("base_url"))
    if provider == "openai":
        return OpenAIChatClient(model=llm_spec.get("model", "gpt-5.4-mini"), base_url=llm_spec.get("base_url"))
    if provider == "stub":
        script = [
            ChatResult(tool=s["tool"], args=s.get("args", {})) if s.get("tool") else ChatResult(text=s.get("text", ""))
            for s in llm_spec.get("script", [{"text": "ok"}])
        ]
        return StubChatClient(script)
    raise ValueError(f"unknown llm provider {provider!r}")


class PipelineSystem:
    """A built, deployed agentic system: a backend runtime + the spec that produced it.

    Implements ``submit(Event) -> TurnResult`` so it can sit behind a ``StreamRuntime``
    (the stream path drives the same CEP-bearing submit)."""

    def __init__(self, spec: Dict[str, Any], backend, graph, tools, retriever, long_term=None, cep=None):
        self.spec = spec
        self.backend = backend
        self.graph = graph
        self.tools = tools
        self.retriever = retriever
        self.long_term = long_term  # optional LongTermStore for resumption + fact archive
        # Declarative CEP rules from the spec's ``cep:`` section (empty if none).
        self.cep = list(cep or [])

    @property
    def backend_name(self) -> str:
        return self.backend.name

    def submit(self, event: Event):
        """Run the turn, then feed the inbound event to every CEP rule (which may fire
        tool/submit actions through the inner backend runtime). CEP runs exactly once per
        inbound event, on this path; submit actions tag their derived events so they can't
        recurse."""
        result = self.backend.submit(event)
        for wiring in self.cep:
            wiring.on_event(event, self.backend, self.tools)
        return result

    def stream(self):
        """Drive a live event stream through this system (CEP fires on the same submit
        path). Returns a ``StreamRuntime`` over this system."""
        from pyagentic.stream import StreamRuntime

        return StreamRuntime(self)


def build_system(spec: Dict[str, Any], backend: Optional[str] = None) -> PipelineSystem:
    """Compile a spec dict and wire it onto the chosen backend (spec ``backend:`` unless
    overridden). A ``stores.conversation`` section hot-swaps the durable store (e.g.
    Redis/Valkey) behind the ConversationStore SPI; ``stores.long_term`` selects the
    resumption + fact archive (memory/postgres) behind the LongTermStore SPI."""
    from pyagentic.cep import compile_cep
    from pyagentic.longterm import make_long_term_store
    from pyagentic.stores import make_conversation_store

    graph, tools, retriever = builder.build(spec, chat_client_factory=_chat_client_factory)
    name = backend or spec.get("backend", "local")
    stores_spec = spec.get("stores") or {}
    store_spec = stores_spec.get("conversation")
    store = make_conversation_store(store_spec) if store_spec else None
    long_term = make_long_term_store(stores_spec.get("long_term")) if stores_spec.get("long_term") else None
    rt = backends.make_backend(name, graph, tools, retriever, store=store)
    cep = compile_cep(spec.get("cep"))
    return PipelineSystem(spec, rt, graph, tools, retriever, long_term=long_term, cep=cep)


def load(path: str, backend: Optional[str] = None) -> PipelineSystem:
    """Load a ``pipeline.yaml`` and build the system."""
    spec = yaml.safe_load(Path(path).read_text(encoding="utf-8"))
    return build_system(spec, backend=backend)
